File: domains/files/listing.py
"""Tool ``list_files``: lista arquivos do workspace."""

from __future__ import annotations

from typing import TYPE_CHECKING, Final, Literal, TypedDict

from core.errors import ErrorPayload, ToolError, guarded

if TYPE_CHECKING:
    from mcp.server.mcpserver import MCPServer

    from domains import Runtime

type FileKind = Literal["video", "audio", "image", "other"]

_KINDS: Final[dict[str, FileKind]] = {
    ".mp4": "video", ".mov": "video", ".mkv": "video", ".webm": "video", ".avi": "video",
    ".mp3": "audio", ".wav": "audio", ".aac": "audio", ".flac": "audio", ".m4a": "audio",
    ".png": "image", ".jpg": "image", ".jpeg": "image", ".webp": "image",
}  # fmt: skip
_MAX_ENTRIES: Final = 500


class FileEntry(TypedDict):
    """Um arquivo do workspace."""

    path: str
    kind: FileKind
    size_bytes: int


class ListFilesResult(TypedDict):
    """Arquivos encontrados."""

    directory: str
    files: list[FileEntry]
    count: int
    truncated: bool


def kind_of(suffix: str) -> FileKind:
    """Classifica um arquivo pela extensão."""
    return _KINDS.get(suffix.lower(), "other")


@guarded
def list_files(
    runtime: Runtime, directory: str = ".", *, kind: FileKind | None = None, recursive: bool = True
) -> ListFilesResult:
    """Implementação pura, testável sem MCP."""
    base = runtime.workspace.resolve(directory)
    if not base.is_dir():
        raise ToolError(f"Diretório '{directory}' não existe.", code="not_found")
    iterator = base.rglob("*") if recursive else base.glob("*")
    entries: list[FileEntry] = []
    truncated = False
    for path in sorted(p for p in iterator if p.is_file()):
        entry_kind = kind_of(path.suffix)
        if kind is not None and entry_kind != kind:
            continue
        if len(entries) >= _MAX_ENTRIES:
            truncated = True
            break
        entries.append(
            FileEntry(
                path=runtime.workspace.relative(path),
                kind=entry_kind,
                size_bytes=path.stat().st_size,
            )
        )
    return ListFilesResult(
        directory=runtime.workspace.relative(base) or ".",
        files=entries,
        count=len(entries),
        truncated=truncated,
    )


def register(mcp: MCPServer[None], runtime: Runtime) -> None:
    """Expõe a tool no servidor."""

    @mcp.tool(name="list_files")
    def _tool(
        directory: str = ".", kind: FileKind | None = None, recursive: bool = True
    ) -> ListFilesResult | ErrorPayload:
        """Lista os arquivos disponíveis no workspace.

        Chame primeiro para descobrir com o que trabalhar. Não altera nada.

        Args:
            directory: Subdiretório do workspace. Raiz por padrão.
            kind: Filtra por tipo: video, audio, image ou other.
            recursive: Inclui subdiretórios.
        """
        return list_files(runtime, directory, kind=kind, recursive=recursive)

File: version-in-rust/domains/files/listing.rs
//! Tool `list_files`: lista arquivos do workspace.

use std::path::{Path, PathBuf};
use std::sync::Arc;

use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use crate::core::errors::{guarded, ErrorCode, ToolError, ToolResult};
use crate::domains::{McpServer, Runtime};

/// Tipo de arquivo, pela extensão.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "lowercase")]
pub enum FileKind {
    Video,
    Audio,
    Image,
    Other,
}

const KINDS: &[(&str, FileKind)] = &[
    (".mp4", FileKind::Video),
    (".mov", FileKind::Video),
    (".mkv", FileKind::Video),
    (".webm", FileKind::Video),
    (".avi", FileKind::Video),
    (".mp3", FileKind::Audio),
    (".wav", FileKind::Audio),
    (".aac", FileKind::Audio),
    (".flac", FileKind::Audio),
    (".m4a", FileKind::Audio),
    (".png", FileKind::Image),
    (".jpg", FileKind::Image),
    (".jpeg", FileKind::Image),
    (".webp", FileKind::Image),
];
const MAX_ENTRIES: usize = 500;

/// Um arquivo do workspace.
#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct FileEntry {
    pub path: String,
    pub kind: FileKind,
    pub size_bytes: u64,
}

/// Arquivos encontrados.
#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct ListFilesResult {
    pub directory: String,
    pub files: Vec<FileEntry>,
    pub count: usize,
    pub truncated: bool,
}

/// Classifica um arquivo pela extensão (com ponto, ex: `.mp4`).
pub fn kind_of(suffix: &str) -> FileKind {
    let lower = suffix.to_ascii_lowercase();
    KINDS
        .iter()
        .find(|(ext, _)| *ext == lower)
        .map_or(FileKind::Other, |(_, kind)| *kind)
}

fn suffix_of(path: &Path) -> String {
    path.extension()
        .map(|ext| format!(".{}", ext.to_string_lossy()))
        .unwrap_or_default()
}

fn collect_files(base: &Path, recursive: bool, out: &mut Vec<PathBuf>) {
    let Ok(entries) = std::fs::read_dir(base) else {
        return;
    };
    for entry in entries.flatten() {
        let path = entry.path();
        if path.is_file() {
            out.push(path);
        } else if recursive && path.is_dir() {
            collect_files(&path, recursive, out);
        }
    }
}

/// Implementação pura, testável sem MCP.
///
/// # Errors
///
/// [`ErrorCode::NotFound`] se o diretório não existir.
pub fn list_files(
    runtime: &Runtime,
    directory: &str,
    kind: Option<FileKind>,
    recursive: bool,
) -> ToolResult<ListFilesResult> {
    let base = runtime.workspace.resolve(directory)?;
    if !base.is_dir() {
        return Err(ToolError::new(
            format!("Diretório '{directory}' não existe."),
            ErrorCode::NotFound,
        ));
    }
    let mut paths = Vec::new();
    collect_files(&base, recursive, &mut paths);
    paths.sort();
    let mut entries = Vec::new();
    let mut truncated = false;
    for path in paths {
        let entry_kind = kind_of(&suffix_of(&path));
        if kind.is_some_and(|wanted| wanted != entry_kind) {
            continue;
        }
        if entries.len() >= MAX_ENTRIES {
            truncated = true;
            break;
        }
        entries.push(FileEntry {
            path: runtime.workspace.relative(&path),
            kind: entry_kind,
            size_bytes: std::fs::metadata(&path).map(|m| m.len()).unwrap_or(0),
        });
    }
    let directory = runtime.workspace.relative(&base);
    Ok(ListFilesResult {
        directory: if directory.is_empty() {
            ".".to_string()
        } else {
            directory
        },
        count: entries.len(),
        files: entries,
        truncated,
    })
}

/// Parâmetros da tool.
#[derive(Debug, Deserialize, JsonSchema)]
pub struct Params {
    /// Subdiretório do workspace. Raiz por padrão.
    #[serde(default = "default_directory")]
    pub directory: String,
    /// Filtra por tipo: video, audio, image ou other.
    #[serde(default)]
    pub kind: Option<FileKind>,
    /// Inclui subdiretórios.
    #[serde(default = "default_true")]
    pub recursive: bool,
}

fn default_directory() -> String {
    ".".to_string()
}

fn default_true() -> bool {
    true
}

/// Expõe a tool no servidor.
pub fn register(mcp: &mut McpServer, runtime: &Arc<Runtime>) {
    let runtime = Arc::clone(runtime);
    mcp.tool(
        "list_files",
        "Lista os arquivos disponíveis no workspace.\n\n\
         Chame primeiro para descobrir com o que trabalhar. Não altera nada.",
        move |params: Params| {
            guarded(list_files(
                &runtime,
                &params.directory,
                params.kind,
                params.recursive,
            ))
        },
    );
}
